File: 백준/Gold/2096. 내려가기/내려가기.java
/**
 * Author: nowalex322, Kim HyeonJae
 */

import java.io.*;
import java.util.*;

public class Main {
    static BufferedReader br;
    static BufferedWriter bw;
    static StringTokenizer st;
    static int N, board[][], maxdp[][], mindp[][];

    public static void main(String[] args) throws Exception {
        new Main().solution();
    }

    public void solution() throws Exception {
        br = new BufferedReader(new InputStreamReader(System.in));
//        br = new BufferedReader(new InputStreamReader(new FileInputStream("src/main/java/BOJ_2096_내려가기/input.txt")));
        bw = new BufferedWriter(new OutputStreamWriter(System.out));

        N = Integer.parseInt(br.readLine());
        board = new int[N][3];
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < 3; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        maxdp = new int[N + 1][3];
        mindp = new int[N + 1][3];
        maxdp[1][0] = mindp[1][0] = board[0][0];
        maxdp[1][1] = mindp[1][1] = board[0][1];
        maxdp[1][2] = mindp[1][2] = board[0][2];

        for (int i = 2; i <= N; i++) {
            maxdp[i][0] = board[i - 1][0] + Math.max(maxdp[i - 1][0], maxdp[i - 1][1]);
            mindp[i][0] = board[i - 1][0] + Math.min(mindp[i - 1][0], mindp[i - 1][1]);
            maxdp[i][1] = board[i - 1][1] + Math.max(maxdp[i - 1][0], Math.max(maxdp[i - 1][1], maxdp[i - 1][2]));
            mindp[i][1] = board[i - 1][1] + Math.min(mindp[i - 1][0], Math.min(mindp[i - 1][1], mindp[i - 1][2]));
            maxdp[i][2] = board[i - 1][2] + Math.max(maxdp[i - 1][1], maxdp[i - 1][2]);
            mindp[i][2] = board[i - 1][2] + Math.min(mindp[i - 1][1], mindp[i - 1][2]);
        }

        int maxRes = Math.max(maxdp[N][0], Math.max(maxdp[N][1], maxdp[N][2]));
        int minRes = Math.min(mindp[N][0], Math.min(mindp[N][1], mindp[N][2]));
        StringBuilder sb = new StringBuilder();
        sb.append(maxRes).append(" ").append(minRes);
        bw.write(sb.toString());
        bw.flush();
        bw.close();
        br.close();
    }
}

File: 백준/Gold/2096. 내려가기/내려가기.cc
/**
 * Author: nowalex322, Kim HyeonJae
 */
#include <bits/stdc++.h>
using namespace std;

// #define int long long
#define MOD 1000000007
#define INF LLONG_MAX
#define ALL(v) v.begin(), v.end()

#ifdef LOCAL
#include "algo/debug.h"
#else
#define debug(...) 42
#endif

void solve() {
    int N;
    cin >> N;

    vector<int> maxdp(3), mindp(3);
    vector<int> nextmaxdp(3), nextmindp(3);

    for (int i = 0; i < 3; i++) {
        int num;
        cin >> num;
        maxdp[i] = mindp[i] = num;
    }

    for (int i = 1; i < N; i++) {
        vector<int> tmp(3);
        for (int j = 0; j < 3; j++) {
            cin >> tmp[j];
        }

        nextmaxdp[0] = tmp[0] + max(maxdp[0], maxdp[1]);
        nextmindp[0] = tmp[0] + min(mindp[0], mindp[1]);
        nextmaxdp[1] = tmp[1] + max(maxdp[0], max(maxdp[1], maxdp[2]));
        nextmindp[1] = tmp[1] + min(mindp[0], min(mindp[1], mindp[2]));
        nextmaxdp[2] = tmp[2] + max(maxdp[1], maxdp[2]);
        nextmindp[2] = tmp[2] + min(mindp[1], mindp[2]);

        maxdp = nextmaxdp;
        mindp = nextmindp;
    }

    int maxres = max(maxdp[0], max(maxdp[1], maxdp[2]));
    int minres = min(mindp[0], min(mindp[1], mindp[2]));

    cout << maxres << " " << minres << "\n";
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int tt = 1;  // 기본적으로 1번의 테스트 케이스를 처리
    // cin >> tt;    // 테스트 케이스 수 입력 (필요 시)

    while (tt--) {
        solve();
    }
    return 0;
}